### PR TITLE
Added Integration Tests and Unit test for Efficient Filtering for Faiss Engine

### PR DESCRIPTION
**Pull request by navneet1v**
### Description
-> Added Integration Tests and Unit test for Efficient Filtering for Faiss Engine.
-> Removed some TODOs and fixed an edge case in the code which failed during Unit tests.

 
### Issues Resolved
#903 
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/test/java/org/opensearch/knn/index/FaissIT.java [hunk 1]
```diff
@@ -280,4 +293,106 @@ public void testEndToEnd_fromModel() throws IOException, InterruptedException {
...
+            .endObject()
+            .endObject()
+            .endObject();
+        final String mapping = Strings.toString(builder);
+
+        createKnnIndex(INDEX_NAME, mapping);
+
+        addKnnDocWithAttributes(
+            DOC_ID,
+            new float[] { 6.0f, 7.9f, 3.1f },
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/knn/index/FaissIT.java:374]: we can generate vector values using random, seems we're not testing exact doc ids in result. This way it's easy to generate more docs and then vary between ann and exact searches

File: src/test/java/org/opensearch/knn/index/FaissIT.java [hunk 2]
```diff
@@ -280,4 +293,106 @@ public void testEndToEnd_fromModel() throws IOException, InterruptedException {
             assertEquals(numDocs - i - 1, Integer.parseInt(results.get(i).getDocId()));
         }
     }
+
+    @SneakyThrows
+    public void testQueryWithFilter_withDifferentCombination_thenSuccess() {
```
Comment (martin-gaievski) [inline src/test/java/org/opensearch/knn/index/FaissIT.java:298]: Should we add different tests for exact and ann searchers? I on a fence if we need it in IT as we already have unit test for this and it might be too low level for IT.

File: src/test/java/org/opensearch/knn/index/query/KNNWeightTests.java [hunk 3]
```diff
@@ -70,6 +78,10 @@ public class KNNWeightTests extends KNNTestCase {
     private static final String CIRCUIT_BREAKER_LIMIT_100KB = "100Kb";
 
     private static final Map<Integer, Float> DOC_ID_TO_SCORES = Map.of(10, 0.4f, 101, 0.05f, 100, 0.8f, 50, 0.52f);
+    private static final Map<Integer, Float> FILTERED_DOC_ID_TO_SCORES = Map.of(101, 0.05f, 100, 0.8f, 50, 0.52f);
```
Comment (jmazanec15) [inline src/test/java/org/opensearch/knn/index/query/KNNWeightTests.java:81]: Maybe add note how these got computed?

File: src/test/java/org/opensearch/knn/index/FaissIT.java (hunk 2 above)
Comment (navneet1v) [reply src/test/java/org/opensearch/knn/index/FaissIT.java:298]: No we should have integration tests. Even when we have Unit tests. The reason why I am combining both is, the setup for the tests are same. Hence I combined them at 1 place. Given that we already have so many integration tests and setting up each integration tests add more time. Hence, using a single setup.

File: src/test/java/org/opensearch/knn/index/FaissIT.java (hunk 1 above)
Comment (navneet1v) [reply src/test/java/org/opensearch/knn/index/FaissIT.java:374]: We can use random generator. But I don't see any point of using random generator. Atleast this way we know what we are ingesting and in future we can compare scores also.

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java [hunk 4]
```diff
@@ -134,12 +137,7 @@ private BitSet getFilteredDocsBitSet(final LeafReaderContext ctx, final Weight f
             return new FixedBitSet(0);
         }
 
-        final BitSet acceptDocs = createBitSet(scorer.iterator(), liveDocs, maxDoc);
-        // TODO: Based on this cost shift to exact search, because even in ANN search you have to calculate the
-        // distance for K vectors. This can avoid calls to native layer and save some latency.
-        final int cost = acceptDocs.cardinality();
-        log.debug("Number of docs valid for filter is = Cost for filtered k-nn is : {}", cost);
-        return acceptDocs;
+        return createBitSet(scorer.iterator(), liveDocs, maxDoc);
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:140]: This was an extra code and doing nothing. I just removed that code.

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java [hunk 5]
```diff
@@ -134,12 +137,7 @@ private BitSet getFilteredDocsBitSet(final LeafReaderContext ctx, final Weight f
             return new FixedBitSet(0);
         }
 
-        final BitSet acceptDocs = createBitSet(scorer.iterator(), liveDocs, maxDoc);
-        // TODO: Based on this cost shift to exact search, because even in ANN search you have to calculate the
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:138]: We are already doing that. This TODO was not removed when I implemented exact search.

File: src/test/java/org/opensearch/knn/index/query/KNNWeightTests.java (hunk 3 above)
Comment (navneet1v) [reply src/test/java/org/opensearch/knn/index/query/KNNWeightTests.java:81]: These were copied from DOC_ID_TO_SCORES which was already present. I removed 1 docId from there.

**Review by martin-gaievski — APPROVED** [review]

File: src/main/java/org/opensearch/knn/index/query/KNNWeight.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (navneet1v) [reply src/main/java/org/opensearch/knn/index/query/KNNWeight.java:140]: added

File: src/test/java/org/opensearch/knn/index/FaissIT.java [hunk 6]
```diff
@@ -44,6 +49,14 @@
 
 public class FaissIT extends KNNRestTestCase {
 
+    private static final String INDEX_NAME = "test-index-1";
+    private static final String FIELD_NAME = "test-field-1";
+    private static final String DOC_ID = "doc1";
```
[the remark this answers is not in the document]
Comment (navneet1v) [reply src/test/java/org/opensearch/knn/index/FaissIT.java:54]: done

**Review by jmazanec15 — APPROVED** [review]
